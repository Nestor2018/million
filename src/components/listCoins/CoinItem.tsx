import React, { useCallback, memo } from 'react';
import { View, Text, StyleSheet, Pressable, Image } from 'react-native';

import colors from '../../utils/colors';
import { Icoin } from '../../interfaces/coin.interface';


interface CoinItemProps {
  coin: Icoin;
  onPress: () => void;
}

/**
 * Componente CoinItem (Renderiza una moneda)
 * @param {CoinItemProps} props - Props que contienen el objeto coin y la función onPress
 * @returns {JSX.Element} El componente renderizado
 */
const CoinItem: React.FC<CoinItemProps> = memo(({ coin, onPress }) => {
  const { symbol, name, price_usd, percent_change_1h } = coin;

  /**
   * Muestra una imagen dependiendo de si el valor de la moneda subió o bajó en la última hora
   * @param {number} percentChange
   * @returns {any} Fuente de la imagen
   */
  const getImageArrow = useCallback((percentChange: number) => {
    return percentChange > 0
      ? require('../../../assets/arrow_up.png')
      : require('../../../assets/arrow_down.png');
  }, []);

  return (
    <Pressable onPress={onPress} style={styles.container} testID="coinItem">
      <View style={styles.row}>
        <Text style={styles.symbolText}>{symbol}</Text>
        <Text style={styles.nameText}>{name}</Text>
        <Text style={styles.price}> {`$${price_usd}`}</Text>
      </View>
      <View style={styles.row}>
        <Text style={styles.percentText}>{percent_change_1h}</Text>
        <Image
          source={getImageArrow(percent_change_1h)}
          style={styles.imageIcon}
        />
      </View>
    </Pressable>
  );
});

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    padding: 16,
    borderBottomColor: colors.zircon,
    borderBottomWidth: 1,
  },
  row: {
    flexDirection: 'row',
  },
  symbolText: {
    color: colors.white,
    fontWeight: 'bold',
    fontSize: 16,
    marginRight: 12,
  },
  nameText: {
    color: colors.white,
    fontSize: 14,
  },
  percentText: {
    color: colors.white,
    fontSize: 12,
    marginRight: 8,
  },
  price: {
    color: colors.white,
    fontSize: 14,
    marginLeft: 12,
  },
  imageIcon: {
    width: 22,
    height: 22,
  },
});

export default CoinItem;
