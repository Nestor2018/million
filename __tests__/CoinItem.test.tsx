import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { Image } from 'react-native';
import CoinItem from '../src/components/listCoins/CoinItem';
import '@testing-library/jest-native/extend-expect';


describe('CoinItem Component', () => {
  const mockCoin = {
    symbol: 'BTC',
    name: 'Bitcoin',
    price_usd: '50000',
    percent_change_1h: 5,
    id: '1',
    nameid: 'bitcoin',
    rank: 1,
    percent_change_24h: 10,
    percent_change_7d: 20,
    market_cap_usd: '1000000000',
    volume24: '500000000',
    volume24a: '450000000',
    csupply: '18000000',
    tsupply: '21000000',
    msupply: '21000000',
  };

  it('should call onPress when pressed', () => {
    const mockOnPress = jest.fn();
    const { getByTestId } = render(
      <CoinItem coin={mockCoin} onPress={mockOnPress} />,
    );

    const coinItem = getByTestId('coinItem');
    fireEvent.press(coinItem);

    expect(mockOnPress).toHaveBeenCalled();
  });

  it('should display the correct coin details', () => {
    const { getByText } = render(
      <CoinItem coin={mockCoin} onPress={() => { }} />,
    );

    expect(getByText('BTC')).toBeTruthy();
    expect(getByText('Bitcoin')).toBeTruthy();
    expect(getByText('$50000')).toBeTruthy();
    expect(getByText('5')).toBeTruthy();
  });

  it('should display the correct arrow icon based on percent change', () => {
    const mockCoinPositive = {
      ...mockCoin,
      percent_change_1h: 5, // Cambio positivo
    };
    const { getByTestId } = render(
      <CoinItem coin={mockCoinPositive} onPress={() => { }} />,
    );
    const image = getByTestId('coinItem').findByType(Image);
    expect(image.props.source).toEqual(require('../assets/arrow_up.png'));

    const mockCoinNegative = {
      ...mockCoin,
      percent_change_1h: -5, // Cambio negativo
    };
    const { getByTestId: getByTestIdNegative } = render(
      <CoinItem coin={mockCoinNegative} onPress={() => { }} />,
    );
    const imageNegative = getByTestIdNegative('coinItem').findByType(Image);
    expect(imageNegative.props.source).toEqual(require('../assets/arrow_down.png'));
  });


  it('should handle missing or invalid data gracefully', () => {
    const mockCoinInvalid = {
      symbol: 'BTC',
      name: 'Bitcoin',
      price_usd: undefined, // Precio no definido
      percent_change_1h: null, // Cambio no definido
      id: '1',
      nameid: 'bitcoin',
      rank: 1,
      percent_change_24h: 10,
      percent_change_7d: 20,
      market_cap_usd: '1000000000',
      volume24: '500000000',
      volume24a: '450000000',
      csupply: '18000000',
      tsupply: '21000000',
      msupply: '21000000',
    };
    const { getByText } = render(
      <CoinItem coin={mockCoinInvalid} onPress={() => { }} />,
    );

    // Verifica que el precio y el cambio porcentual se muestren correctamente
    expect(getByText('$undefined')).toBeTruthy(); // Precio vacÃ­o
  });

  it('should not re-render when props do not change', () => {
    const mockOnPress = jest.fn();
    const { rerender, getByTestId } = render(
      <CoinItem coin={mockCoin} onPress={mockOnPress} />,
    );

    const initialRender = getByTestId('coinItem');
    expect(initialRender).toBeTruthy();

    // Re-renderiza el componente con las mismas props
    rerender(<CoinItem coin={mockCoin} onPress={mockOnPress} />);
    const afterRerender = getByTestId('coinItem');

    // Verifica que el componente no haya cambiado
    expect(afterRerender).toBe(initialRender);
  });

  it('should match snapshot', () => {
    const tree = render(<CoinItem coin={mockCoin} onPress={() => { }} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

