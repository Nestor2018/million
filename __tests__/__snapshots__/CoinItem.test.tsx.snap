// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinItem Component should match snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="coinItem"
>
  <View>
    <Text>
      BTC
    </Text>
    <Text>
      Bitcoin
    </Text>
    <Text>
       
      $50000
    </Text>
  </View>
  <View>
    <Text>
      5
    </Text>
    <Image
      source={
        {
          "testUri": "../../../assets/arrow_up.png",
        }
      }
    />
  </View>
</View>
`;
